subsys/fb/cfb.c

```diff
@@ -84,103 +137,406 @@ static inline uint8_t get_glyph_byte(uint8_t *glyph_ptr, const struct cfb_font *
 }
 
 /*
- * Draw the monochrome character in the monochrome tiled framebuffer,
- * a byte is interpreted as 8 pixels ordered vertically among each other.
+ * Transferring buffer contents to display
+ *
+ * @param param buffer transfer param
+ * @param x X position
+ * @param y X position
+ * @param w width
+ * @param h height
+ *
+ * @return negative value if failed, otherwise 0
  */
-static uint8_t draw_char_vtmono(const struct char_framebuffer *fb,
-				char c, uint16_t x, uint16_t y,
-				bool draw_bg)
+static int transfer_buffer(const struct transfer_param *param, int16_t x, int16_t y, int16_t w,
+			   int16_t h)
 {
-	const struct cfb_font *fptr = &(fb->fonts[fb->font_idx]);
-	const bool need_reverse = (((fb->screen_info & SCREEN_INFO_MONO_MSB_FIRST) != 0)
-			     != ((fptr->caps & CFB_FONT_MSB_FIRST) != 0));
-	uint8_t *glyph_ptr;
+	const struct display_driver_api *api = param->dev->api;
+	struct display_buffer_descriptor desc;
+	int err = 0;
+
+	if (param->offscreen) {
+		return 0;
+	}
 
-	if (c < fptr->first_char || c > fptr->last_char) {
-		c = ' ';
+	if (x >= param->fb->x_res || y >= param->fb->y_res) {
+		return 0;
 	}
 
-	glyph_ptr = get_glyph_ptr(fptr, c);
-	if (!glyph_ptr) {
+	if ((y + h) < 0) {
 		return 0;
 	}
 
-	for (size_t g_x = 0; g_x < fptr->width; g_x++) {
-		const int16_t fb_x = x + g_x;
+	if (y < 0) {
+		h += y;
+		y = 0;
+	}
 
-		for (size_t g_y = 0; g_y < fptr->height; g_y++) {
-			/*
-			 * Process glyph rendering in the y direction
-			 * by separating per 8-line boundaries.
+	desc.buf_size = param->size;
+	desc.width = w;
+	desc.height = h;
+	desc.pitch = w;
+
+	if (desc.height + y >= param->fb->y_res) {
+		desc.height = param->fb->y_res - y;
+	}
+
+	if (desc.width + x >= param->fb->x_res) {
+		desc.width = param->fb->x_res - x;
+		desc.pitch = param->fb->x_res - x;
+	}
+
+	err = api->write(param->dev, x, y, &desc, param->buf);
```

Comment:
Why is the display driver API `write` function being called directly from the API pointer here? What is the advantage of this approach over `display_write(param->dev, x, y, &desc, param->bus)`?

Reply:
No reason to not use `display_write`.
Fixed it.
It came from existing source code.
I also fixed similar problems.